R/weight.R

```diff
@@ -41,33 +43,216 @@ weight <- function (x) {
+#'
+#' Change which variables can be set as a dataset's weight.
+#'
+#' @param x a dataset or variable catalog
+#' @param vars Variables to add or remove  this can be a numeric Crunch variable,
+#' list of numeric Crunch variables or a character vector with the aliases of numeric Crunch variables.
+#' @param type a character string determining how the weightVariables
+#' will be modified:
+#' - `"append"` : add `vars` to the current weight variables
+#' - `"remove"` : remove `vars` from the current list of weight variables
+#' - `"replace"`: replace the current weight variables with `vars`
+#'
+#' @export
+modifyWeightVariables <- function(x, vars, type = "append") {
+        current <- crGET(shojiURL(x, "orders", "weights"))
+    new <- current
+    if (is.null(vars)) {
+        new$graph <- NULL
+        type = "remove"
+    } else {
+        if (is.variable(vars) || (length(vars) == 1) & !is.character(vars)) {
+            vars <- list(vars)
+        }
+        if (is.character(vars)) {
+            ds <- loadDataset(datasetReference(self(x)))
+            vars <- lapply(vars, function(v){
+                if (v %in% names(x)) {
+                    ds[[v]]
+                } else {
+                    v
+                }
+            })
+        }
+        all_var <- vapply(vars, is.Numeric, logical(1))
+        if (!all(all_var)) {
+            var_names <- vapply(vars, function(v){
+                if (is.variable(v)) {
+                    return(name(v))
+                } else {
+                    return(as.character(v))
+                }
+            }, character(1))
+            err_text <- " is not a numeric Crunch variable."
+            if ( sum(!all_var) > 1) {
+                err_text <- " are not a numeric Crunch variables."
+            }
+            halt( serialPaste(var_names[!all_var]), err_text)
+        }
+        if (type == "append") {
+            new$graph <- unique(c( current$graph, lapply(vars, self)))
+        } else if (type == "remove") {
+            new$graph <- setdiff( current$graph, lapply(vars, self))
+        } else if (type == "replace") {
+            new$graph <- lapply(vars, self)
+        }
+    }
+    if (!identical(current, new)) {
+        crPUT(shojiURL(x, "orders", "weights"),
+            body = toJSON(new))
+        x <- refresh(x)
+    }
+    return(x)
+}
+
+#' @rdname weightVariables
+#' @export
+is.weightVariable <- function(x){
+    if (is.variable(x)) {
+        ds <- loadDataset(datasetReference(self(x)))
+        return(alias(x) %in% weightVariables(ds))
+    } else {
+        return(FALSE)
+    }
+}
+
+#' @rdname weightVariables
+#' @export
+setMethod("is.weightVariable<-", "NumericVariable", function(x, value) {
+    if (!(is.logical(value)  && length(value) == 1)) {
+        halt("Right hand side must be TRUE or FALSE.")
+    }
+    ds <- loadDataset(datasetReference(self(x)))
+    varcat <- allVariables(ds)
+    if (value) {
+        modifyWeightVariables(varcat, x, "append")
```

Comment:
FWIW this could be one line with `ifelse(value, "append", "remove")`, or you could pull that ifelse out to its own line like `mode <- ifelse(value, "append", "remove")` and then pass that to `modifyWeightVariables`